### PR TITLE
Convert Zipkin receiver and exporter to use OTLP and fix translation bugs

### LINKED ISSUE #526: Zipkin to Jaeger Span translation error flag not set
**Problem** : JaegerUI isnt highlighting spans with error status when zipkin traces are received by otelcol and exported as jaeger spans.
**Why is this happening ?**
JaegerUI depends on Jaeger Span Model : Where "error" tag of "Bool" type is set by the jaeger collector upon receiving zipkin span with error status code. 

Issue : OtelCol translation from zipkin model to jaeger model sets error tag as string and not as bool  during the translation.

**Input [Zipkin Span]**
  {
    "traceId": "631ef3c9f9250805",
    "id": "631ef3c9f9250805",
    "kind": "SERVER",
    "name": "http:/one",
    "timestamp": 1580830219488000,
    "duration": 29312,
    "localEndpoint": {
      "serviceName": "foo-service",
      "ipv4": "192.168.1.8",
      "port": 9001
    },
    "tags": {
      **_"error": "Request processing failed; nested exception is org.springframework.web.client.HttpServerErrorException: 500 null",_**
      "http.host": "localhost",
      "http.method": "GET",
      "http.path": "/one",
      "http.status_code": "200",
      "http.url": "http://localhost:9001/one",
      "mvc.controller.class": "FooController",
      "mvc.controller.method": "one",
      "spring.instance_id": "192.168.1.8:order-service-sleuth:9001"
    }
  }
 
**Actual [Otel Translated Jaeger Span - where error tag is still string]**
{
    "traceIdLow": 6186857723629179000,
    "traceIdHigh": 0,
    "spanId": 6186857723629179000,
    "parentSpanId": 0,
    "operationName": "http:/one",
    "flags": 0,
    "startTime": 1580830930191000,
    "duration": 108645,
    "tags": [
      {
        "key": "http.status_code",
        "vType": "STRING",
        "vStr": "200"
      },
      {
        "key": "spring.instance_id",
        "vType": "STRING",
        "vStr": "192.168.1.8:order-service-sleuth:9001"
      },
      {
        "key": "http.method",
        "vType": "STRING",
        "vStr": "GET"
      },
      {
        "key": "http.path",
        "vType": "STRING",
        "vStr": "/one"
      },
      {
        "key": "port",
        "vType": "STRING",
        "vStr": "9001"
      },
      {
        "key": "http.url",
        "vType": "STRING",
        "vStr": "http://localhost:9001/one"
      },
      {
        "key": "mvc.controller.class",
        "vType": "STRING",
        "vStr": "FooController"
      },
      {
        "key": "mvc.controller.method",
        "vType": "STRING",
        "vStr": "one"
      },
      **{
        "key": "error",
        "vType": "STRING",
        "vStr": "Request processing failed; nested exception is org.springframework.web.client.HttpServerErrorException: 500 null"
      }**,
      {
        "key": "http.host",
        "vType": "STRING",
        "vStr": "localhost"
      },
      {
        "key": "ipv4",
        "vType": "STRING",
        "vStr": "192.168.1.8"
      },
      {
        "key": "span.kind",
        "vType": "STRING",
        "vStr": "server"
      },
      {
        "key": "status.code",
        "vType": "LONG",
        "vLong": 2
      }
    ]
  }

**Expected [JagerSpan where error type is changed to bool and its value is set as errorMessage]**
{
    "traceIdLow": 8380789575664951000,
    "traceIdHigh": 0,
    "spanId": 8380789575664951000,
    "parentSpanId": 0,
    "operationName": "http:/one",
    "flags": 0,
    "startTime": 1580830714594000,
    "duration": 79701,
    "tags": [
      {
        "key": "http.url",
        "vType": "STRING",
        "vStr": "http://localhost:9001/one"
      },
      {
        "key": "mvc.controller.class",
        "vType": "STRING",
        "vStr": "FooController"
      },
      {
        "key": "mvc.controller.method",
        "vType": "STRING",
        "vStr": "one"
      },
      **{
        "key": "error",
        "vType": "BOOL",
        "vBool": true
      },
      {
        "key": "errorMessage",
        "vType": "STRING",
        "vStr": "Request processing failed; nested exception is org.springframework.web.client.HttpServerErrorException: 500 null"
      },**
      {
        "key": "http.method",
        "vType": "STRING",
        "vStr": "GET"
      },
      {
        "key": "http.host",
        "vType": "STRING",
        "vStr": "localhost"
      },
      {
        "key": "http.path",
        "vType": "STRING",
        "vStr": "/one"
      },
      {
        "key": "ipv4",
        "vType": "STRING",
        "vStr": "192.168.1.8"
      },
      {
        "key": "port",
        "vType": "STRING",
        "vStr": "9001"
      },
      {
        "key": "spring.instance_id",
        "vType": "STRING",
        "vStr": "192.168.1.8:order-service-sleuth:9001"
      },
      {
        "key": "http.status_code",
        "vType": "STRING",
        "vStr": "200"
      },
      {
        "key": "span.kind",
        "vType": "STRING",
        "vStr": "server"
      },
      {
        "key": "status.code",
        "vType": "LONG",
        "vLong": 2
      }
    ]
  }

### LINKED ISSUE #572: Zipkin Post from Envoy 1.13.0 to otel-collector sometime fails because of timestamp value format
Envoy 1.13 now only supports zipkin v2 api but sometimes send an invalid payload related to the timestamp using this format: 1.58266276796031e+15. There is already an open issue on Envoy side https://github.com/envoyproxy/envoy/issues/9341 with discussion on Go side to support these kind of value or not... Can this case be handled in otel-collector? Could wait for a long time for Envoy fix...

**Request:**
POST /api/v2/spans HTTP/1.1
host: tracing_zipkin_cluster
content-type: application/json
x-envoy-internal: true
x-forwarded-for: 10.1.72.4
x-envoy-expected-rq-timeout-ms: 5000
transfer-encoding: chunked
[{"duration":2741,"kind":"SERVER","localEndpoint":{"port":0,"serviceName":"echocall1","ipv4":"10.1.72.4"},"id":"3691a8bd8b50c194","tags":{"upstream_cluster":"h1_ingress_cluster","guid:x-request-id":"e7761137-f1a0-9cae-a4b4-b7c1855d7c52","http.protocol":"HTTP/1.1","component":"proxy","http.url":"http://echocall1.testmgmt.devappdirect.me/echo","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36","downstream_cluster":"-","request_size":"0","response_flags":"-","response_size":"11","http.status_code":"200","http.method":"GET"},"shared":true,"traceId":"3691a8bd8b50c194","name":"something","timestamp":1.58266276796031e+15}]

**Response:**
HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
X-Content-Type-Options: nosniff
Date: Tue, 25 Feb 2020 20:32:51 GMT
Content-Length: 98
json: cannot unmarshal number 1.58266276796031e+15 into Go struct field .timestamp of type uint64

### LINKED ISSUE #1138: Span translation to Zipkin data loss
**Describe the bug**
Translation of OTLP protocol data by Zipkin exporter results in some data loss and incorrect translations.

**Steps to reproduce**
Execute golden dataset correctness tests and view test results summary.

**What did you expect to see?**
Zero failure count.

**What did you see instead?**
Failure count of 9283 and failures: Attributes,0!=2,count=3586; Links,2!=0,count=2441; Status,&v1.Status{Code:0, Message:""}!=(*v1.Status)(nil),count=241; TraceState,"lasterror=f39cd56cc44274fd5abd07ef1164246d10ce2955"!="",count=2719; Kind,1!=0,count=296;

**What version did you use?**
Version: latest

**What config did you use?**
Config: Correctness test config per receiver-exporter combination.

**Environment**
OS: any
Compiler(if manually compiled): any

**Additional context**
Recommend fixing #1136 first as this will resolve some errors and then rerun correctness tests to determine what issues remain.

### LINKED ISSUE #1303: Migrate Zipkin receiver to the new OTLP-based internal data model
**Is your feature request related to a problem? Please describe.**
Zipkin receiver uses old OpenCensus data model which causes unnecessary translation: internal data Zipkin -> OC -> internal data model.

**Describe the solution you'd like**
Zipkin receiver should use OTLP-based internal data model (consumer/pdata/trace.go) instead of OC.

### LINKED ISSUE #1304: Migrate Zipkin exporter to the new OTLP-based internal data model
**Is your feature request related to a problem? Please describe.**
Zipkin exporter uses old OpenCensus data model which causes unnecessary translation: internal data model -> OC -> Zipkin.

**Describe the solution you'd like**
Zipkin exporter should use OTLP-based internal data model (consumer/pdata/trace.go) instead of OC.

### PR DESCRIPTION
**Pull request by kbrockhoff**
**Description:** New translators were created for Zipkin formats to OTLP and for OTLP to Zipkin v2. The Zipkin receiver and exporter were converted to use the new translators.

**Link to tracking Issue:**
Fixes #526
Fixes #572 
Fixes #1138
Fixes #1303
Fixes #1304

**Testing:** All existing tests pass

**Documentation:** N/A

### CODE REVIEW

File: translator/trace/zipkin/zipkinv1_thrift_to_traces.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+func V1ThriftBatchToInternalTraces(zSpans []*zipkincore.Span) ([]pdata.Traces, error) {
+	ocTraces, err := V1ThriftBatchToOCProto(zSpans)
+	if err != nil {
+		return make([]pdata.Traces, 0), err
+	}
+
+	traces := make([]pdata.Traces, 0, len(ocTraces))
+	for _, td := range ocTraces {
+		traces = append(traces, internaldata.OCToTraceData(td))
+	}
```
Comment (bogdandrutu) [inline translator/trace/zipkin/zipkinv1_thrift_to_traces.go:33]:
```golang
td := pdata.NewTraces()
for _, td := range ocTraces {
	tmp := internaldata.OCToTraceData(td)
	tmp.ResourceSpans().MoveAndAppendTo(td.ResourceSpans())
}
```

File: translator/trace/zipkin/zipkinv1_to_traces.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	ocTraces, err := V1JSONBatchToOCProto(blob)
+	if err != nil {
+		return make([]pdata.Traces, 0), err
+	}
+
+	traces := make([]pdata.Traces, 0, len(ocTraces))
+	for _, td := range ocTraces {
+		traces = append(traces, internaldata.OCToTraceData(td))
+	}
+	return traces, nil
```
Comment (bogdandrutu) [inline translator/trace/zipkin/zipkinv1_to_traces.go:32]: Same comment to have only one `pdata.Traces`.

**Comment by bogdandrutu** [comment]
Seems that correctness tests for Zipkin translation are failing

**Comment by kbrockhoff** [comment]
Timestamps are not converting correctly on Linux but are on OS X. I am working on debugging.

**Comment by bogdandrutu** [comment]
Can be some time zone issue, it is just a guess

File: receiver/zipkinreceiver/trace_receiver.go [hunk 3]
```diff
@@ -175,10 +174,10 @@ func (zr *ZipkinReceiver) v2ToTraceSpans(blob []byte, hdr http.Header) (reqs []c
 	}
 
 	if err != nil {
-		return nil, err
+		return pdata.Traces{}, err
 	}
 
-	return zipkin.V2BatchToOCProto(zipkinSpans)
```
Comment (chris-smith-zocdoc) [inline receiver/zipkinreceiver/trace_receiver.go:181]: Can the old method `V2BatchToOCProto` be removed?

File: translator/trace/zipkin/zipkinv2_to_traces.go [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+	}
+	diff = strings.Compare(extractInstrumentationLibrary(b[i]), extractInstrumentationLibrary(b[j]))
+	return diff <= 0
+}
+
+func (b byOTLPTypes) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+// V2SpansToInternalTraces translates Zipkin v2 spans into internal trace data.
```
Comment (chris-smith-zocdoc) [inline translator/trace/zipkin/zipkinv2_to_traces.go:94]: It doesn't appear that this method shares any implementation details with the Zipkin v1, can we keep them somewhat unified? I had just merged them together recently https://github.com/open-telemetry/opentelemetry-collector/pull/1002

File: translator/trace/zipkin/zipkinv2_to_traces.go [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+	attrs.InitEmptyWithCapacity(len(tags))
+	if err := zTagsToInternalAttrs(zspan, tags, attrs); err != nil {
+		return err
+	}
+
+	err := populateSpanEvents(zspan, dest.Events())
+	return err
+}
+
+func populateSpanStatus(tags map[string]string, status pdata.SpanStatus) {
```
Comment (chris-smith-zocdoc) [inline translator/trace/zipkin/zipkinv2_to_traces.go:187]: The previous implementation used [statusMapper](https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/zipkin/status_code.go#L51) to infer the status from additional tags like `http.status_code` https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/zipkin/zipkinv2_to_protospan.go#L239-L252

File: translator/trace/zipkin/zipkinv2_to_traces.go [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+			dest.InsertString(conventions.AttributeNetPeerIP, zspan.RemoteEndpoint.IPv6.String())
+		}
+		if zspan.RemoteEndpoint.Port > 0 {
+			dest.UpsertInt(conventions.AttributeNetPeerPort, int64(zspan.RemoteEndpoint.Port))
+		}
+	}
+	return parseErr
+}
+
+func tagsToAttributeMap(tags map[string]string, dest pdata.AttributeMap) error {
```
Comment (chris-smith-zocdoc) [inline translator/trace/zipkin/zipkinv2_to_traces.go:358]: This logic is different than the previous parsing logic. I'd appreciate if this remained unified with v1 also.

https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/zipkin/zipkinv1_to_protospan.go#L253-L266

I don't think ints should need to be in the `intAttributes` to be parsed. 

Support for bools was dropped, it should be added back.

File: receiver/zipkinreceiver/factory_test.go [hunk 7]
```diff
@@ -36,7 +36,7 @@ func TestCreateDefaultConfig(t *testing.T) {
 type mockTraceConsumer struct {
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply receiver/zipkinreceiver/factory_test.go]: Converted

File: receiver/zipkinreceiver/proto_parse_test.go [hunk 8]
```diff
@@ -107,88 +107,177 @@ func TestConvertSpansToTraceSpans_protobuf(t *testing.T) {
...
+							},
 						},
 					},
 				},
 			},
 		},
-	}
+	})
 
-	assert.Equal(t, want, reqs)
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply receiver/zipkinreceiver/proto_parse_test.go]: Added more testing

File: receiver/zipkinreceiver/trace_receiver.go [hunk 9]
```diff
@@ -116,16 +115,16 @@ func (zr *ZipkinReceiver) Start(ctx context.Context, host component.Host) error
 }
 
 // v1ToTraceSpans parses Zipkin v1 JSON traces and converts them to OpenCensus Proto spans.
-func (zr *ZipkinReceiver) v1ToTraceSpans(blob []byte, hdr http.Header) (reqs []consumerdata.TraceData, err error) {
+func (zr *ZipkinReceiver) v1ToTraceSpans(blob []byte, hdr http.Header) (reqs []pdata.Traces, err error) {
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply receiver/zipkinreceiver/trace_receiver.go]: Changed

File: translator/trace/zipkin/zipkinv1_thrift_to_traces.go (hunk 1 above)
Comment (kbrockhoff) [reply translator/trace/zipkin/zipkinv1_thrift_to_traces.go]: Changed

File: translator/trace/zipkin/zipkinv1_to_traces.go (hunk 2 above)
Comment (kbrockhoff) [reply translator/trace/zipkin/zipkinv1_to_traces.go]: Changed

File: receiver/zipkinreceiver/trace_receiver.go (hunk 3 above)
Comment (kbrockhoff) [reply receiver/zipkinreceiver/trace_receiver.go:182]: Deleted

File: translator/trace/zipkin/zipkinv2_to_traces.go (hunk 5 above)
Comment (kbrockhoff) [reply translator/trace/zipkin/zipkinv2_to_traces.go:190]: Status is going to be removed from the specification. We should wait until that happens to revise this.

File: translator/trace/zipkin/zipkinv2_to_traces.go (hunk 6 above)
Comment (kbrockhoff) [reply translator/trace/zipkin/zipkinv2_to_traces.go:363]: Changed to use same logic

File: testbed/testbed/receivers.go [hunk 10]
```diff
@@ -254,9 +254,8 @@ func (zr *ZipkinDataReceiver) Start(tc *MockTraceConsumer, _ *MockMetricConsumer
 	cfg := factory.CreateDefaultConfig().(*zipkinreceiver.Config)
 	cfg.SetName(zr.ProtocolName())
 	cfg.Endpoint = fmt.Sprintf("localhost:%d", zr.Port)
-
-	params := component.ReceiverCreateParams{Logger: zap.NewNop()}
```
Comment (bogdandrutu) [inline testbed/testbed/receivers.go:258]: All changes in this file can be reverted.

File: exporter/zipkinexporter/zipkin_test.go [hunk 11]
```diff
@@ -65,15 +67,16 @@ func TestZipkinExporter_roundtripJSON(t *testing.T) {
...
 	assert.NoError(t, err)
-	require.NotNil(t, zexp)
+	require.NotNil(t, tes)
 
 	// The test requires the spans from zipkinSpansJSONJavaLibrary to be sent in a single batch, use
 	// a mock to ensure that this happens as intended.
 	mzr := newMockZipkinReporter(cst.URL)
 
 	// Run the Zipkin receiver to "receive spans upload from a client application"
+	zexp := processor.NewTraceFanOutConnector([]consumer.TraceConsumer{tes})
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply exporter/zipkinexporter/zipkin_test.go]: reverted

File: exporter/zipkinexporter/factory_test.go [hunk 12]
```diff
@@ -26,24 +26,28 @@ import (
 )
 
 func TestCreateDefaultConfig(t *testing.T) {
-	cfg := createDefaultConfig()
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply exporter/zipkinexporter/factory_test.go]: reverted

File: exporter/zipkinexporter/zipkin.go [hunk 13]
```diff
@@ -83,39 +81,30 @@ func createZipkinExporter(cfg *Config) (*zipkinExporter, error) {
...
-	tbatch := make([]*zipkinmodel.SpanModel, 0, numSpans)
-	for _, octd := range octds {
-		for _, span := range octd.Spans {
-			zs, err := zipkin.OCSpanProtoToZipkin(octd.Node, octd.Resource, span, ze.defaultServiceName)
-			if err != nil {
-				return numSpans, consumererror.Permanent(fmt.Errorf("failed to push trace data via Zipkin exporter: %w", err))
-			}
-			tbatch = append(tbatch, zs)
-		}
+func (ze *zipkinExporter) PushTraceData(ctx context.Context, td pdata.Traces) (int, error) {
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply exporter/zipkinexporter/zipkin.go]: reverted

File: exporter/zipkinexporter/zipkin_test.go [hunk 14]
```diff
@@ -314,6 +326,7 @@ func TestZipkinExporter_roundtripProto(t *testing.T) {
 	mzr.serializer = zipkinproto.SpanSerializer{}
 
 	// Run the Zipkin receiver to "receive spans upload from a client application"
+	zexp := processor.NewTraceFanOutConnector([]consumer.TraceConsumer{tes})
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply exporter/zipkinexporter/zipkin_test.go]: reverted

File: exporter/zipkinexporter/test_utils.go [hunk 15]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2020, OpenTelemetry Authors
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply exporter/zipkinexporter/test_utils.go]: reverted

File: exporter/zipkinexporter/factory.go [hunk 16]
```diff
@@ -36,14 +36,15 @@ const (
 	defaultServiceName string = "<missing service name>"
 )
 
-// NewFactory creates a factory for OTLP exporter.
+// NewFactory creates a factory for Zipkin exporter
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply exporter/zipkinexporter/factory.go]: reverted

File: exporter/zipkinexporter/zipkin_test.go [hunk 17]
```diff
@@ -151,7 +162,7 @@ type mockZipkinReporter struct {
 	serializer zipkinreporter.SpanSerializer
 }
 
-var _ zipkinreporter.Reporter = (*mockZipkinReporter)(nil)
+var _ (zipkinreporter.Reporter) = (*mockZipkinReporter)(nil)
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply exporter/zipkinexporter/zipkin_test.go:162]: reverted

File: receiver/zipkinreceiver/factory_test.go [hunk 18]
```diff
@@ -23,31 +23,31 @@ import (
...
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/config/configcheck"
+	"go.opentelemetry.io/collector/config/configerror"
 	"go.opentelemetry.io/collector/exporter/exportertest"
 )
 
 func TestCreateDefaultConfig(t *testing.T) {
-	cfg := createDefaultConfig()
+	factory := NewFactory()
+	cfg := factory.CreateDefaultConfig()
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply receiver/zipkinreceiver/factory_test.go]: reverted

File: testbed/testbed/receivers.go (hunk 10 above)
Comment (kbrockhoff) [reply testbed/testbed/receivers.go]: reverted

File: receiver/zipkinreceiver/trace_receiver_test.go [hunk 19]
```diff
@@ -30,34 +30,26 @@ import (
...
 	"go.opentelemetry.io/collector/exporter/zipkinexporter"
-	"go.opentelemetry.io/collector/internal"
 	"go.opentelemetry.io/collector/testutil"
-	"go.opentelemetry.io/collector/translator/internaldata"
-	"go.opentelemetry.io/collector/translator/trace/zipkin"
+	"go.opentelemetry.io/collector/translator/conventions"
 )
 
-const zipkinReceiverName = "zipkin_receiver_test"
+const zipkinReceiver = "zipkin_receiver_test"
```
[the remark this answers is not in the document]
Comment (kbrockhoff) [reply receiver/zipkinreceiver/trace_receiver_test.go]: reverted this variable name

**Pull request merged**